### PR TITLE
[RFC] Retry fgets on EINTR (#7632)

### LINKED ISSUE #7632: Calls to fgets should be retried when error is EINTR
- `nvim --version`: NVIM v0.2.2
- Vim behaves differently? Not sure.
- Operating system/version: macOS / 10.12.6 (16G29)
- Terminal name/version: Terminal.app / Version 2.7.3 (388.1.1)
- `$TERM`: xterm-256color-it-cs (256 color with italics and cursor shape)

The calls to `fgets` in `src/nvim/if_cscope.c` can fail if a signal is delivered during any of the system calls it makes. In my particular case, a SIGCHLD is *always* interfering with calls into `fgets`. The caller should retry when NULL is returned with `errno` set to EINTR.

I don't know how to reproduce this without a configuration. Here's how I noticed it:

1. Install [fzy](https://github.com/jhawthorn/fzy) and the plugin [vim-picker](https://github.com/srstevenson/vim-picker).
2. Generate a cscope database for a C project and `:cscope add` it to nvim. Verify that cscope is hooked up with `:cscope show`.
3. Route `:tag` requests through cscope first with `:set cscopetag` and `:set cscopetagorder=0`.
4. Use the vim-picker plugin's buffer tag search with `:PickerBufferTag`.
5. Select a known tag to jump to in the current file.

I expected to jump to that tag, but instead, I got a generic cscope communication error:

```
Error detected while processing function 11[5]..10:
line    4:
E262: error reading cscope connection 0
```

I built a debug version of neovim, put a breakpoint on `cs_reading_emsg`, and noticed that `errno` was set to EINTR. I then set lldb to print when signals occur with `process handle --notify true --pass true`, and saw only SIGCHLD signals getting delivered, likely since fzy is exiting after making the selection.

I hit this every time I select a tag using vim-picker. I'm not sure if or when this regressed, or if it's an issue in vim.

### PR DESCRIPTION
**Pull request by mattwidmann**
This fixes #7632 for me. It's tricky to test, since it relies on a race between a subprocess exiting and waiting for cscope's output.

- Should we instead change all of these to use `vim_fgets`? I didn't here because `vim_fgets` has some additional handling for lines that are too long, but that might be alright for all or some of these cases.
- The linter is complaining about missing braces on if statements I didn't change -- is that expected?

### CODE REVIEW

**Comment by justinmk** [comment]
Thanks!

> Should we instead change all of these to use vim_fgets

Better to change one thing at a time.

> The linter is complaining about missing braces on if statements I didn't change -- is that expected?

Yes, it "cascades" to nearby lines, so legacy code gets updated incrementally.

File: src/nvim/fileio.c [hunk 1]
```diff
@@ -4448,14 +4448,24 @@ bool vim_fgets(char_u *buf, int size, FILE *fp) FUNC_ATTR_NONNULL_ALL
...
     buf[size - 1] = NUL;            /* Truncate the line */
 
     /* Now throw away the rest of the line: */
     do {
       tbuf[FGETS_SIZE - 2] = NUL;
+retry_ignore:
+      errno = 0;
       ignoredp = fgets((char *)tbuf, FGETS_SIZE, fp);
+      if (ignoredp == NULL && errno == EINTR) {
+        goto retry_ignore;
```
Comment (justinmk) [inline src/nvim/fileio.c]: technically we're not ignoring the retval now, so `ignoredp` should not be used.

**Comment by justinmk** [comment]
would be good to include your explanation and debugging steps from https://github.com/neovim/neovim/issues/7632 , in the commit message.

File: src/nvim/fileio.c (hunk 1 above)
Comment (mattwidmann) [reply src/nvim/fileio.c]: Yeah, I guess we can just rely on the existing loop to retry the `fgets` here.

Comment (mattwidmann) [reply src/nvim/fileio.c]: Wait, this won't properly handle an EOF, either...

Comment (justinmk) [reply src/nvim/fileio.c]: just to be clear, i was only speaking about the specific `ignoredp` global which is used as a dummy assignment when Vim doesn't care about some return value.

**Comment by mattwidmann** [comment]
I had to rewrite the `vim_fgets` function to make it safe (could just spin forever if EOF is returned from one of the calls to `fgets`. I could put this in a separate commit when I squash, since it's a separate (though overlapping) fix.

I also pushed a linter fix and an empty commit with the proposed commit message.

**Comment by justinmk** [comment]
LGTM, just leave a comment when ready.

**Comment by mattwidmann** [comment]
Looking at the test failure here: <http://neovim-qb.szakmeister.net/build/15079/tap_report/by_test?reportset=functionaltest-freebsd-64-debug>

Is this test flaky? This change could have caused issues here, but the test looks somewhat racy. The failure is:

```
test/functional/eval/timer_spec.lua @ 41
Failure message: test/functional/eval/timer_spec.lua:44: Expected objects to be the same.
Passed in:
(number) 1
Expected:
(number) 0
stack traceback:
test/functional/eval/timer_spec.lua:44: in function <test/functional/eval/timer_spec.lua:41>

test/functional/eval/timer_spec.lua @ 49: timers works with zero timeout
```

And the test is:

```lua
  it('are triggered during sleep', function()
    command("call timer_start(50, 'MyHandler', {'repeat': 2})")
    nvim_async("command", "sleep 10")
    eq(0,eval("g:val"))
    run(nil, nil, nil, 300)
    eq(2,eval("g:val"))
  end)
```

There's nothing preventing the `timer_start` command and the check that `g:val` equals 0 from being separated by 50 ms or more. This didn't reproduce on the other test harnesses, either.

**Comment by justinmk** [comment]
Yes, that test is flaky. Can ignore it here.

**Comment by mattwidmann** [comment]
Alright -- I've pushed two commits here and tested them locally. One of the fileio tests is failing on 32-bit:

```
[32m[2m[ RUN      ][0m[0m file_read can read small chunks of input until eof: 
```

That's an alarming test to have fail against this branch. I didn't change that fileio code -- it uses its own buffering scheme instead of `FILE` streams, so it makes the read or readv syscalls itself.

Have you seen that test fail before? Is there any way I can test 32-bit on macOS? I ran that test on my machine and it passed, though it took about a second to complete. Is it timing out, maybe?

**Comment by justinmk** [comment]
@mattwidmann That is a very weird failure, it looks like a timeout (build log stopped abruptly), but travis didn't flag it as a timeout. Have not seen that before.

I restarted the test since the logs don't show anything useful,  to see if the failure is consistent.

**Comment by mattwidmann** [comment]
Looks like the build failed this time, trying to fetch dependencies:

```E: Failed to fetch http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu/pool/main/g/gcc-5/g++-5-multilib_5.4.1-2ubuntu1~14.04_amd64.deb  Unable to connect to ppa.launchpad.net:http:
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
apt-get.diagnostics
apt-get install failed
[...]
The command "sudo -E apt-get -yq --no-install-suggests --no-install-recommends --force-yes install autoconf automake apport build-essential clang-4.0 cmake cscope g++-5-multilib g++-multilib gcc-5-multilib gcc-multilib gdb language-pack-tr libc6-dev-i386 libtool llvm-4.0-dev locales pkg-config unzip valgrind xclip" failed and exited with 100 during .
```

Might need to be restarted again.

**Comment by justinmk** [comment]
restarted again. definitely looks like travis issue:

```
oci runtime error: exec failed: container_linux.go:265: starting container process caused "could not create session key: disk quota exceeded"
```

**Comment by mattwidmann** [comment]
Well, I'm done with this PR -- feel free to merge it when you're ready. Let me know if anything should be changed.

**Pull request merged**